README

CDI2Nero 0.9.9 - (C) 2002 by DeXT

http://cdirip.cjb.net


*** WARNING! This is beta software. It has not been fully tested and may
*** contain bugs. Use it at you own risk.

*** This tool must not be used to duplicate copyright-protected content.
*** The author will not be responsible for its improper usage.


Purpose
-------

This little tool will convert a CDI (DiscJuggler) image to NRG (Nero) format.
Both CDI 2.0 and 3.0 image versions are supported.

It currently has support for:

 - CD-ROM/XA type CDs
 - Mode 2 and Audio tracks
 - Single session and Multisession images

It lacks support for:

 - Audio-only CDs (TAO only)
 - Mode 1 data tracks

These will be added soon.


Usage
-----

Usage is very simple, just double-clic the executable file and an Open File
dialog will appear, select your source CDI image to convert and press OK.
A Save File dialog will then appear on the same path as source image. Write
the desired name for the destination NRG image (you don't need to write any
extension) and press OK. Default name is "image.nrg".

You can also use it from command-line. Syntax is:

cdi2nero image.cdi [image.nrg]

For command-line usage, if you don't specify a destination filename, the
source image filename will be used but with NRG extension.


Burning
-------

Once converted you can burn the newly created image. If a DAO warning message
appears, you should select Yes (although this depends on burner, I think).
This will only appear on Multisession images.


Notes
-----

This is a unified DAO/TAO version, unlike previous ones, where you can select
NRG image type when saving. DAO images are most standard ones, and is the only
format that Nero itself is able to generate. TAO images are special ones,
which lacks a cuesheet, and track positions are managed by the burner itself.
Nero is not capable of generating these. On most burners, these images will
be burned in "Track at Once" mode (hence the name) instead of default DAO.

A general note is to burn in DAO (default) mode and if something fails, try
TAO instead. But note that if your burner doesn't support "short" tracks (i.e.
those below 302 sectors) you still will get errors when trying to burn such
images. These errors are usually "cannot write track at correct position",
or seeing CD-Extra discs as Audio-only (i.e. no data track).

Below is a list of tested burners, reported by users, which get successful or
failed in burning process:

Success:

Acer CRW 6206A
BTC BCE-432IE
Gateway 8x4x32
HP CD-Writer 7100i
HP CD-Writer 7200e
HP CD-Writer 8100i
HP CD-Writer 9100i
HP CD-Writer 9310i
Iomega Zip CD 650
Iomega Zip CD USB (DAO only)
Memorex CRW-1622 (DAO only)
Mitsumi 4801TE
OTI-975 Socrates
Philips CDD3610
Philips CDD3801
Philips CDD4201
Philips PCRW404K
RICOH MP9060A
Sony CRX-140E
Teac CD-R56S
Yamaha 16x

Failure:

HP CD-Writer 7200i (*)
HP CD-Writer 8210e
HP CD-Writer 8250i (*)
Memorex tri-max 200
Mitsumi 4802TE
Mitsumi 4804TE
Sony CRX-100E
Sony CRX-120
Sony CRX-145E
TDK VeloCD 12x

(*) works with non-short tracks, patched or fixed images (tested)

Please note that these "problematic" burners mostly have "short" track
problems, and they should still work fine with non "short" track images. This
has not been tested, unless specified with (*).

You can identify "short" track type images with following CDIrip command-line:

cdirip image.cdi /info >info.txt

Open info.txt and search for tracks below 302 sectors in size, if any.

You can try to fix them with CDIfix before converting to NRG (only in case of
multiple audio track images).


That's all. Hope you like it!

Many thanks to xeal and myn, who have been excellent beta-testers!


History
-------

 0.9.9
 
  - Support for DiscJuggler 4.0 CDI images
  - cdi2nero now opens CDR-WIN 5.x (*.cdr) images, too
  - Unified code with CDIrip

 0.9b8

  - Added support for CDI 3.5.826+ images

 0.9b7

  - Fixed bug in 0.9b6 with some old (3.0, 2.0) images (were not recognized)

 0.9b6

  - Fixed bug with some DJ 3.5 images (only 1st session was converted)

 0.9b5
 
  - Added new DiscJuggler 3.0.780 CDI image support
 
 0.9b4 (unified)
 
  - Unified TAO/DAO version in a single executable
  - Support for raw data tracks
  - New buffering system (great speedup!)
  - Major internal cleanup
  - New compiler (Bloodshed Dev-C++ v4.0) - tiny executable size!

 0.9b3 (DAO)

  - Fixed multiple audio track bug
  - New progress counter (faster, bigger and better :)

 0.9b2 (TAO)

  - New TAO-type NRG image
  - Fixed multiple audio track bug
  - Audio-only CD support (well, not really - this derives from TAO images)

 0.9b

  - First release
  - Buggy :(


(C) 2000 DeXT


ChangeLog

CDI2Nero 0.9.9 - (C) 2002 by DeXT

Modifications by SiZ! on 14 nov 06
http://sbibuilder.shorturl.com/

Changes :
- added a TAO "-t" option for command line
- added a "-h" help option

usage (on command line) :
cdi2nero <input.cdi> [output.nrg] [-t]

examples :

cdi2nero input.cdi
-> generates a DAO "input.nrg" image

cdi2nero input.cdi output.nrg
-> generates a DAO "output.nrg" image

cdi2nero input.cdi -t
-> generates a TAO "input.nrg" image

cdi2nero input.cdi output.nrg -t
-> generates a TAO "output.nrg" image

src/common.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#ifdef _WIN32
#include <windows.h>
#endif
#include "common.h"


long fread_as_little (void *buf, long length, FILE *fsource)
{
     char *tmp;
     long i;

     tmp = buf;

#ifdef BIG_ENDIAN
     for (i = length; i > 0; i--)
         if (!fread(&(tmp[i-1]), 1, 1, fsource)) return 0; // error
#else
     for (i = 0; i < length; i++)
         if (!fread(&(tmp[i]), 1, 1, fsource)) return 0; // error
#endif

return 1; // success
}

long fread_as_big (void *buf, long length, FILE *fsource)
{
     char *tmp;
     long i;

     tmp = buf;

#ifdef BIG_ENDIAN
     for (i = 0; i < length; i++)
         if (!fread(&(tmp[i]), 1, 1, fsource)) return 0; // error
#else
     for (i = length; i > 0; i--)
         if (!fread(&(tmp[i-1]), 1, 1, fsource)) return 0; // error
#endif

return 1; // success
}

long fwrite_as_little (void *buf, long length, FILE *fdest)
{
     char *tmp;
     long i;

     tmp = buf;

#ifdef BIG_ENDIAN
     for (i = length; i > 0; i--)
         if (!fwrite(&(tmp[i-1]), 1, 1, fdest)) return 0; // error
#else
     for (i = 0; i < length; i++)
         if (!fwrite(&(tmp[i]), 1, 1, fdest)) return 0; // error
#endif

return 1; // success
}

long fwrite_as_big (void *buf, long length, FILE *fdest)
{
     char *tmp;
     long i;

     tmp = buf;

#ifdef BIG_ENDIAN
     for (i = 0; i < length; i++)
         if (!fwrite(&(tmp[i]), 1, 1, fdest)) return 0; // error
#else
     for (i = length; i > 0; i--)
         if (!fwrite(&(tmp[i-1]), 1, 1, fdest)) return 0; // error
#endif

return 1; // success
}


void error_exit(long errcode, char *string)
{

char errmessage[256];

#ifdef _WIN32
HWND hWnd = NULL;
#endif

    if (errcode != ERR_GENERIC)
       printf("\n%s: %s\n", string, strerror(errno));  // string is used as Filename

    switch(errcode)
          {
          case ERR_OPENIMAGE:
                   strcpy(errmessage, "File not found\n\nYou may have typed a wrong name or path to source CDI image");
                   break;
          case ERR_SAVETRACK:
                   strcpy(errmessage, "Could not save track");
                   break;
          case ERR_READIMAGE:
                   strcpy(errmessage, "Error reading image");
                   break;
          case ERR_PATH:
                   strcpy(errmessage, "Could not find destination path");
                   break;
          case ERR_GENERIC:
          default: strcpy(errmessage, string);          // string is used as Error message
          }

#ifdef _WIN32
    MessageBoxA(hWnd, errmessage, NULL, MB_OK | MB_ICONERROR);
    ExitProcess(0);
#else
    printf(errmessage);
    exit(EXIT_FAILURE);
#endif

}


src/common.h
#ifndef __COMMON_H__
#define __COMMON_H__

/* Error codes */

#define ERR_GENERIC   0L
#define ERR_OPENIMAGE 0x01
#define ERR_SAVETRACK 0x02
#define ERR_READIMAGE 0x03
#define ERR_SAVEIMAGE 0x04
#define ERR_PATH      0x05


/* For Debug only! */

// #define DEBUG_CDI



/* Basic structures */

typedef struct image_s
       {
       long               header_offset;
       long               header_position;
       long               length;
       unsigned long      version;
       unsigned short int sessions;
       unsigned short int tracks;
       unsigned short int remaining_sessions;
       unsigned short int remaining_tracks;
       unsigned short int global_current_session;
       } image_s;

typedef struct track_s
       {
       unsigned short int global_current_track;
       unsigned short int number;
       long               position;
       unsigned long      mode;
       unsigned long      sector_size;
       unsigned long      sector_size_value;
       long               length;
       long               pregap_length;
       long               total_length;
       unsigned long      start_lba;
       unsigned char      filename_length;
       } track_s;


/* Functions */

long fread_as_little(void *buf, long length, FILE *fsource);
long fread_as_big(void *buf, long length, FILE *fsource);
long fwrite_as_little(void *buf, long length, FILE *fdest);
long fwrite_as_big(void *buf, long length, FILE *fdest);

void error_exit(long errcode, char *string);

#endif


src/cdi2nero.h
#ifndef __CDI2NERO_H__
#define __CDI2NERO_H__

/* Track format */

#define FORMAT_DAOI 0x01
#define FORMAT_ETNF 0x02

/* Save format */

#define DAO       1
#define TAO       2
#define CANCELLED 0xFF

/* Buffers */

#define READ_BUF_SIZE  1024*1024
#define WRITE_BUF_SIZE 1024*1024

/* Structures */

typedef struct opts_s
       {
       unsigned char ask_source_image;
       unsigned char ask_dest_image;
       unsigned char track_format;
       unsigned char image_format;
       unsigned char do_cut;
       unsigned char do_convert;
       } opts_s;

/* Functions */

void savetrack(FILE *fsource, FILE *fdest, track_s *track, image_s *image, opts_s *opts);

unsigned long lba_to_msf(unsigned long lba);
unsigned long bcd_conv(unsigned long value);

unsigned char askfilename(char *string);
unsigned char asksavefilename(char *string);

void mem_write(void *data, long size, int *index, char *buffer);
void mem_write_as_big(void *data, long size, int *index, char *buffer);
void mem_write_int_big(long value, long size, int *index, char *buffer);

void NRG_write_cues_hdr (char *fcues, int *fcues_i, image_s *image, unsigned long track_type);
void NRG_write_daoi_hdr (char *fdaoi, int *fdaoi_i, image_s *image);
void NRG_write_cues_track (char *fcues, int *fcues_i, track_s *track);
void NRG_write_daoi_track (char *fdaoi, int *fdaoi_i, track_s *track, long nrg_offset);
void NRG_write_cues_tail (char *fcues, int *fcues_i, track_s *track);
void NRG_write_sinf (char *fdaoi, int *fdaoi_i, image_s *image);
void NRG_write_etnf_hdr (char *fdaoi, int *fdaoi_i, image_s *image);
void NRG_write_etnf_track (char *fdaoi, int *fdaoi_i, track_s *track, opts_s *opts, long nrg_offset);

#endif



src/cdi.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include "common.h"
#include "cdi.h"

// Global variables

unsigned long temp_value;


/////////////////////////////////////////////////////////////////////////////


unsigned long ask_type(FILE *fsource, long header_position)
{

unsigned char filename_length;
unsigned long track_mode;

    fseek(fsource, header_position, SEEK_SET);
    fread_as_little(&temp_value, 4, fsource);
    if (temp_value != 0)
       fseek(fsource, 8, SEEK_CUR); // extra data (DJ 3.00.780 and up)
    fseek(fsource, 24, SEEK_CUR);
    fread_as_little(&filename_length, 1, fsource);
    fseek(fsource, filename_length, SEEK_CUR);
    fseek(fsource, 19, SEEK_CUR);
    fread_as_little(&temp_value, 4, fsource);
       if (temp_value == 0x80000000)
          fseek(fsource, 8, SEEK_CUR); // DJ4
    fseek(fsource, 16, SEEK_CUR);
    fread_as_little(&track_mode, 4, fsource);
    fseek(fsource, header_position, SEEK_SET);
    return (track_mode);
}


/////////////////////////////////////////////////////////////////////////////


void CDI_read_track (FILE *fsource, image_s *image, track_s *track)
{

     char TRACK_START_MARK[10] = { 0, 0, 0x01, 0, 0, 0, 0xFF, 0xFF, 0xFF, 0xFF };
     char current_start_mark[10];

         fread_as_little(&temp_value, 4, fsource);
         if (temp_value != 0)
            fseek(fsource, 8, SEEK_CUR); // extra data (DJ 3.00.780 and up)

         fread(&current_start_mark, 10, 1, fsource);
         if (memcmp(TRACK_START_MARK, current_start_mark, 10)) error_exit(ERR_GENERIC, "Unsupported format: Could not find the track start mark");

         fread(&current_start_mark, 10, 1, fsource);
         if (memcmp(TRACK_START_MARK, current_start_mark, 10)) error_exit(ERR_GENERIC, "Unsupported format: Could not find the track start mark");

         fseek(fsource, 4, SEEK_CUR);
         fread_as_little(&track->filename_length, 1, fsource);
         fseek(fsource, track->filename_length, SEEK_CUR);
         fseek(fsource, 11, SEEK_CUR);
         fseek(fsource, 4, SEEK_CUR);
         fseek(fsource, 4, SEEK_CUR);
         fread_as_little(&temp_value, 4, fsource);
            if (temp_value == 0x80000000)
               fseek(fsource, 8, SEEK_CUR); // DJ4
         fseek(fsource, 2, SEEK_CUR);
         fread_as_little(&track->pregap_length, 4, fsource);
         fread_as_little(&track->length, 4, fsource);
         fseek(fsource, 6, SEEK_CUR);
         fread_as_little(&track->mode, 4, fsource);
         fseek(fsource, 12, SEEK_CUR);
         fread_as_little(&track->start_lba, 4, fsource);
         fread_as_little(&track->total_length, 4, fsource);
         fseek(fsource, 16, SEEK_CUR);
         fread_as_little(&track->sector_size_value, 4, fsource);

         switch(track->sector_size_value)
               {
               case 0 : track->sector_size = 2048; break;
               case 1 : track->sector_size = 2336; break;
               case 2 : track->sector_size = 2352; break;
               default: error_exit(ERR_GENERIC, "Unsupported sector size");
               }

         if (track->mode > 2) error_exit(ERR_GENERIC, "Unsupported format: Track mode not supported");

         fseek(fsource, 29, SEEK_CUR);
         if (image->version != CDI_V2)
            {
            fseek(fsource, 5, SEEK_CUR);
            fread_as_little(&temp_value, 4, fsource);
            if (temp_value == 0xffffffff)
                fseek(fsource, 78, SEEK_CUR); // extra data (DJ 3.00.780 and up)
            }
}


void CDI_skip_next_session (FILE *fsource, image_s *image)
{
     fseek(fsource, 4, SEEK_CUR);
     fseek(fsource, 8, SEEK_CUR);
     if (image->version != CDI_V2) fseek(fsource, 1, SEEK_CUR);
}

void CDI_get_tracks (FILE *fsource, image_s *image)
{
     fread_as_little(&image->tracks, 2, fsource);
}

void CDI_init (FILE *fsource, image_s *image, char *fsourcename)
{
     fseek(fsource, 0L, SEEK_END);
     image->length = ftell(fsource);

     if (image->length < 8) error_exit(ERR_GENERIC, "Image file is too short");

     fseek(fsource, image->length-8, SEEK_SET);
     fread_as_little(&image->version, 4, fsource);
     fread_as_little(&image->header_offset, 4, fsource);

     if (errno != 0) error_exit(ERR_READIMAGE, fsourcename);
     if (image->header_offset == 0) error_exit(ERR_GENERIC, "Bad image format");
}

void CDI_get_sessions (FILE *fsource, image_s *image)
{
#ifndef DEBUG_CDI
     if (image->version == CDI_V35)
        fseek(fsource, (image->length - image->header_offset), SEEK_SET);
     else
        fseek(fsource, image->header_offset, SEEK_SET);

#else
     fseek(fsource, 0L, SEEK_SET);
#endif
     fread_as_little(&image->sessions, 2, fsource);
}

